README.md
Caribon
=======

A repetition detector written in Rust.

[![Build Status](https://travis-ci.org/lise-henry/caribon.svg?branch=master)](https://travis-ci.org/lise-henry/caribon)

Demo server
===========

If you want to have an idea of what Caribon does without downloading
anything, you can have a look at
[this instance that runs Caribon as a web service](http://vps184889.ovh.net/caribon/).

Downloading
===========

Either use `git` to get the latest version:

`$ git clone https://github.com/lise-henry/caribon.git`

or just download one of the stable(ish)
[releases](https://github.com/lise-henry/caribon/releases).

(If you only plan to use Caribon as a library for your rust program,
you don't need to worry too much about downloading and building, just
add `caribon = "*"` in your `Cargo.toml` file.)


Build
=====

You'll need Rust and Cargo, see their [install instructions](http://www.rust-lang.org/install.html). Then

`$ cargo build`

should do the job, though you can also run `cargo build --release`
directly. You can then run caribon either with:

`$ cargo run --release`

or by directly executing the binary (in `target/debug` or
`target/release`):

`$ target/release/caribon`

Installing
==========

If you have a recent version of `Cargo`, you can use:

`$ cargo install caribon`

(for the latest version on [crates.io](http://crates.io), or

`$ cargo install --git https://github.com/lise-henry/caribon`

for the latest version on GitHub)

which will download, build and install Caribon.

Cargo run
=========

If you don't want to install Caribon, `cargo run` might be the
simplest option. Note, though, that command-line arguments must
be prefixed by `--` so cargo gives them to the binary: 

`$ cargo run -- --input=some_text.txt`

Also note that, by default, `cargo run` builds and runs the program in
debug mode, which is slower. This isn't a problem for tiny files, but
if you plan to detect a repetitions in, say, a novel, using
cpu-extensive features (such as fuzzy string matching, see below), you might want
to run with `--release`:

`$ cargo run --release -- --input=big_file.html --output=output_big_file.html`

Examples
========

Here is an
[example](https://lise-henry.github.io/rust/caribon-examples/example_readme.html)
of Caribon using the HTML output of a (previous) version of this 
README, obtained with the following command:

`$ caribon --language=english --input=README.html
--output=example.html --fuzzy=0.5`

(Note that `--fuzzy=0.5`, while useful to show that fuzzy string
matching does indeed work, is not a very sensible parameter as is it
quite high (words only needs to be 50% similar to be considered the
same, matching e.g. `just` and `rust`). For real life usage, a lower value
would be recommended.)

Here is another [example](https://lise-henry.github.io/rust/caribon-examples/screenshot.png), displaying repetitions in
`README.md` to the terminal, using the following command:

`$ caribon --language=english --input=README.md --fuzzy=0.5 | more`

![example](https://lise-henry.github.io/rust/caribon-examples/screenshot.png)

While outputting to the terminal might be useful for small files, HTML
outputs gives a more useful result, as higlighting a word will show
you the other occurrences of it.

Options
=======

Caribon provides a list of options. Here's the
explanations to a few ones, from the most commons the the pretty
advanced ones:

### Language ###

* `--language=[english|french|spanish|...]`specifies the language of the
  input file. It is important for two reasons. The first one is that
  Caribon internally uses a stemming library, which will detect when
  words are derived from the same stem, e.g. "eats", "eat" and
  "eating" will be considered the same word. (More information on how
  this stemming library works can be found on the
  [Snowball project website](http://snowball.tartarus.org/).) The
  second reason is that for some languages (currently only french and
  english), Caribon provides a default list of words to ignore for
  repetition counting (e.g. in english "it", "a" and so on are on it)
  to avoid cluttering the result file. It is possible to disable
  stemming by using "no_stemmer" instead of a language. This isn't
  really advised, but it might be useful if you want to try Caribon on
  a language that isn't implemented.
* `--list-languages` prints the list of languages supported by the
  stemming library.

### Input and output ###

* `--input=[file]` specifies the input file. By default it is `stdin`,
which means you'll have directly to type your text and end it with
`control-D`. If `file` is a non-existing file, the program aborts.
* `--output=[file]` specifies the output file. It defaults to `stdout`,
printing the result to the terminal.

The input and output filenames extension determine the input and
output format, e.g. if you pass `--input=text.html --output=result.html`, Caribon will
infer that the content is in HTML and that it must also output HTML
(so `$ caribon < input.html > output.html` is NOT equivalent to `$
caribon --input=input.html --output=output.html`: in the first case,
Caribon will consider the input as raw text and will output in
`terminal` format (see below), while in the latter one it will
understand that both files are HTML).

It is possible to override this behaviour by specifying

* `--input-format=[text|html]` or
* `--output-format=[terminal|html|markdown]`.

A note on the `terminal` output format: it is designed to print text
to the terminal, by underlining and colouring some words with UNIX
terminal special characters (see screenshot above). It is, thus, only activated when no
output file name is given and Caribon prints on the standard output,
HTML output being the default for most of the cases.

### Text statistics ###

* `--print-stats`, if passed to Caribon, will also display some statistics
  on the input text on the standard output.

### Threshold and max-distance ###

The most useful algorithm of Caribon is local repetition
detections. It detects when a word is repeated in a given interval of
words. This interval is determined by

* `--max-distance=[value]` (default is currently 50).

So basically, if `max-distance` is 50 and the word 'foo' occurs twice in this
interval, each occurrence will have a "repetition value" of 2. If
'foo' is repeated a third time in a 50-words interval *after the second
occurence*, then each of these occurrences will have a repetition
value of 3. (If there is then more than 50 words without apparition of
'foo', and 'foo' appears again, the value of the latest apparition
will be reset to 1).

Words are underlined when their "repetition value" is higher than a
threshold, which can be set by:

* `--threshold=[value]`. The default is `1.9`, so a word will be underlined
  as soon that is is repeated two times locally. If you change the threshold to, say,
  `2.5`, a word will have to be repeated three times (locally) to be
  underlined.

(Why a float value for the threshold, instead of an integer one?
Because the local repetition detector will underline words in
different colors: green, orange and red according to the "severity" of
the repetitions. So setting the threshold to `1.01` or `1.99` will not
change which words are underlined, but they will be in orange or red
more quickly in the first case.)

### Fuzzy string matching ###

Caribon uses a stemming library to detect words that are part of the
same 'family'. It turns out that this algorithm is not always
enough, and particularly it doesn't detect repetitions when there is a
typo (e.g. "higlight" and "highlight" should probably be considered a
repetition, even if it is mispelled in the first case). To solve this
issue, there is the option of activating fuzzy string matching:

* `fuzzy=[value]`, where the value is a number between 0.0 and 1.0 which
  represents the maximal 'difference' between two words until they are
  no more identical: a value of 0.2 means that two words must be at
  most "20% different" until they are no more considered the same.

Internally, this algorithm uses the
[Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)
(and more specifically the
[Rust implementation by Florian Ebelling](https://crates.io/crates/edit-distance))
which computes a distance between two strings by estimating the number
of insertions, deletions and permutations it require to go from one to
another. E.g., "dog" and "dogs" have a distance of 1, while "dog" and
"cat" have a distance of 3. This value is then divided by the length
of the string to match, and two string are considered "identical" (or,
at least, a repetition) when this value is less than the value given
to `--fuzzy=`.

E.g., with `--fuzzy=0.2`, "highlight" and "higlight" will have a
"difference" estimated to 1/9 (Levenshtein distance of 1, it only needs
one deletion to go from the first to the second, divided by the length
of "highlight", 9), so it will be a repetition. "Just" and "Rust" will
have a "difference" of 1/4, so won't be considered a repetition.

Fuzzy matching is practical, but you should not set it to a too high
value, else you will have a lot of false positives. Empirically, `0.2`
or `0.25` is a good choice.

Fuzzy matching has a drawback: it requires a lot more of CPU. Caribon
still manages to run reasonably fast (e.g., less than a second to
detect repetitions on a whole novel, with fuzzy string matching
activated) but it only uses fuzzy string matching for local
repetitions, and not for global ones (see below).

### Global repetitions ###

By default, Caribon only detects repetitions at a local level (if they
are separated by less than `max-distance` words). It is,
however, possible to activate global repetition detecting with:

* `--global-threshold=[value]`, value being (again) a number between 0.0
and 1.0.

In this case, a word will be considered a repetition (even if it is
never repeated in a `max-distance` range of words) if the relative
number of occurence is higher than the global threshold. I.e., if
`global-threshold` is set to 0.01, a word will be highlighted (in
blue) if it represents more than 1% of the total number of words in
the document.

### Ignored words ###

Some words, like "a" or "the", are unavoidably repeated a
lot and it doesn't make much sense to consider them a repetition. It
is thus useful to ignore some words. `Caribon` provides a 
default list for english and french, but it is in all cases possible
to provide your own with:

* `--ignore="list of common words"`.

This list must be separated by either spaces or commas (or, actually,
anything that isn't a letter), and must be encircled by
quotes. This list *replaces* the default one
provided by Caribon (for english and french, at least). If you want to
*add* words to these list instead of replace it, use:

* `--add-ignored="list of more ignored words"`

Another option for ignoring words is:

* `--ignore-proper=[true|false]` (default is to false)

If sets to true, Caribon will try to ignore proper nouns". That is, a word will not
count for repetition if it starts with a capital letter and
is not at the beginning of a sentence.

Library
=======

It is possible to use Caribon as a Rust library. The documentation is
available [here](http://lise-henry.github.io/rust/caribon/index.html); in order to
be certain to have the documentation version corresponding to the code
you downloaded, you can also generate it with
`cargo doc`.

Caribon library is also available on
[Crates.io](https://crates.io/crates/caribon), allowing you to easily
use it in any Cargo project: just add

`caribon = "0.7"`

in the dependencies section of your
`Cargo.toml` file.

Caribon-server
==============

If you are not a big fan of command-line interface, you can have a
look at [Caribon-server](https://github.com/lise-henry/caribon-server) that
runs Caribon as a web service (using
[Iron framework](https://github.com/iron/iron)). See [here for an
instance running it](http://vps184889.ovh.net/caribon/).

Current features
================

* Built-in list of ignored words (common words whose repetitions don't
  matter) for french and english, though they are not complete.
* Stemming support for languages supported by the Snowball (http://snowball.tartarus.org/)
  project.
* Additionally (because stemming algorithms aren't always perfect, and sometimes
  you make typos), support for fuzzy string matching (based on Levenhstein distance).
* Count repetitions locally and globally.
* Detects HTML tags in input. Normally works both for HTML fragments
  or full HTML pages.
* Outputs the detected repetitions either in an HTML file (the most
  useful option), directly to the terminal, or to a Markdown file (with less useful information).

ChangeLog
=========

[See here](ChangeLog.md).

License
=======

Caribon is licensed under the
[GNU Lesser General Public License](LICENSE), version 2.1
or (at your convenience) any ulterior version.

Credits
=======

Caribon is written by Élisabeth Henry `<liz.henry at ouvaton.org>`.

This software uses (rust bindings to) the
[C Stemming library](http://snowball.tartarus.org/dist/libstemmer_c.tgz)
written by Dr Martin Porter, licensed under the BSD License.

It also uses the [Rust implementation](https://crates.io/crates/edit-distance) of
Levenshtein distance written by Florian Ebelling, licensed under the Apache 2.0 License.

ToDo 
====

Library
-------
* Make colour highlighting more configurable
* Complete builtin lists of ignored words and provide them for other
  languages (currently, only french, and english);
* Provide algorithm to detect repetitions of expressions, not just
  single words;
* Make library callable from C (and other languages than Rust);
* Enhance documentation and add tests.

Program
-------
* Add options to select highligting colours
* Find better default values?
* Make different repositories for program and library?
* Add a variant with GUI (Gtk+?)?



ChangeLog.md
ChangeLog
=========

0.8.1 (2017-03-04)
----------------------
* Update `strsim` dependency to 0.6.0

0.8.0 (2017-01-07)
----------------------
* Add `ast_to_ispell` method to parser to allow Caribon to be runned
pretending to be ispell.
* Add `ast_to_repetitions` method to parser to make Caribon more usable as a library.

0.7.4 (2016-10-26)
----------------------
* Use `strsim` instead of `edit_distance` crate; benchmark shows that
  it increases the performances for fuzzy matching.

0.7.3 (2016-09-30)
------------------
* Use only the first two words to detect the ignored list to use, so
  it works if the user uses a unicode language code
* Add some words to french ignored list


0.7.2 (2016-08-17)
------------------
* Fix a formatting bug when printing HTML from a text input
* Runned `clippy` on repository and fixed some unesthetic stuff

0.7.1 (2015-12-10)
------------------
* Fixed documentation, homepage and repository links because of change
  of username

0.7.0 (2015-11-12)
------------------
* In `ast_to_*`, use `string.push_str(...)` instead of `string =
  string + ...`
* Replaced `to_string()` by `to_owned()` to convert `&str` to `String`
* Moved to version 0.3 of stemmer library, requiring a bit more
functions to be `&mut self`
* Fixed `Cargo.toml` to no longer use `*` dependencies which broke
  everything (sorry -_-)


0.6.2 (2015-07-15)
------------------
* Moved library files and binary files into separate directories
* Changed license to GNU Lesser General Public License
* Binary: added `--print-stats` option (a bit limited at the moment)

0.6.1 (2015-07-13)
------------------
* `Parser.detect_global` now only does something if there is enough
  words in the file compared to the threshold

0.6.0 (2015-07-08)
------------------
* Added `Parser.with_more_ignored` to add ignored words instead of replacing them
* Added `--add-ignored=` option for the binary
* Caribon now supports working on full HTML files, not just HTML
  fragments. This necessitated some changes in the API:
   * Use a struct `Ast` instead of `Vec<Word>`
   * Renamed `words_to_*` to `ast_to_*`
   * `detect_[local|global]` methods now takes a mutable ref to an Ast
   instead of moving a `Vec<Word>` and returning a new one
* Removed `detect_leak` algorithm
* Added special "no_stemmer" option to `Parser::new`: disables
stemming
* Fixed a bug in fuzzy string matching



0.5.2 (2015-07-03)
------------------
* Added `Parser.words_to_terminal` that outputs repetitions for the
terminal instead of using HTML, using terminal colours characters
* Added `Parser.words_to_markdown` that outputs repetitions in
markdown
* Added fuzzy string matching possibility, enabled by constructing the
parser with the `with_fuzzy` method
* Reworked options for the binary:
  * Added `--fuzzy=` to trigger fuzzy string matching
  * Disabled global detection by default
  * Added `--input-format` and `--output-format` options
  * Without these options, auto-detect input and output formats based
    on file extensions

0.5.1 (2015-07-01)
------------------
* Bugfixes in `Parser.tokenize` and `Parser.words_to_html` when
  handling non-html input text

0.5.0 (2015-07-01)
------------------
* Improved documentation
* `Parser.tokenize` now returns a `Result`, giving an `Error` instead of
panicking
* `Parser::new` also returns a `Result` instead of an `Option`
* `detect_leak` now uses `max_distance` to compute a leak, so `leak`
parameter is removed.
* Therefore, `--leak=x` argument is removed from the binary
* Options for the binary have some more changes (see `--help`)
* It is now possible to chain detection methods, highlighting
with different colours
* Added `words_stats` method



0.4.0 (2015-06-30)
------------------
* `words_to_html` is now part of Parser implementation
* If `parser.html` is false (input type is not html),
  `parser.words_to_html` convert line breaks to `<br/>`
* Escape characters such as `&nbsp;` are leaved as such if
  `parser.html` is true
* Better support for user-defined list of ignored words in the library
* New option `--ignore=` in the binary
  

0.3.0 (2015-06-29)
------------------
* `words_to_html` takes another argument, `standalone`
* removed "--debug" option
* added "--input" and "--output" options

0.2.2 (2015-06-28)
------------------
* Fixed a bug (and unnecessary clone) in detect_local algorithm
* Added a list of ignored words for english

0.2.1 (2015-06-28)
------------------
* Added documentation for the library and examples on Github.io
* Minor bugfixes

0.2.0 (2015-06-28)
------------------
* Added support for HTML in input
* Added some ignored words for french
* Added option to ignore proper nouns
* Added `detect_local` algorithm
* Added `Parser::list_languages()` function
* Added support for command-line arguments

0.1.0 (2015-06-27)
------------------

* first release


src/config.rs
// This file is part of Caribon.
//
// Caribon is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// Caribon is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Caribon.  If not, see <http://www.gnu.org/licenses/>.

use std::env;
use std::process::exit;
use std::fs::File;
use std::error::Error;
use std::io;
use std::io::Read;
use std::io::Write;
use caribon::Parser;

static ARG_LANG:&'static str = "--language=";
static ARG_THRESHOLD:&'static str = "--threshold=";
static ARG_MAX_DISTANCE:&'static str = "--max-distance=";
static ARG_GLOBAL_THRESHOLD:&'static str = "--global-threshold=";
static ARG_INPUT_FORMAT:&'static str = "--input-format=";
static ARG_OUTPUT_FORMAT:&'static str = "--output-format=";
static ARG_IGNORE_PROPER:&'static str = "--ignore-proper=";
static ARG_USAGE:&'static str = "--help";
static ARG_INPUT:&'static str = "--input=";
static ARG_OUTPUT:&'static str = "--output=";
static ARG_VERSION:&'static str = "--version";
static ARG_LIST_LANGUAGES:&'static str = "--list-languages";
static ARG_IGNORE:&'static str = "--ignore=";
static ARG_ADD_IGNORED:&'static str = "--add-ignored=";
static ARG_FUZZY:&'static str = "--fuzzy=";

pub fn list_languages() {
    println!("Supported languages:");
    for l in Parser::list_languages() {
        println!("- '{}'", l);
    }
}

fn version() {
    println!("{}", env!("CARGO_PKG_VERSION"));
}

fn usage() {
    println!("
Caribon, version {} by Élisabeth Henry <liz.henry@ouvaton.org>

Detects the repetitions in a text and highlights them

Options:
  {}: displays this message
  {}: displays program version
  {}: lists the implemented languages
  {}[language]: sets the language of the text (default: french)
  {}[filename]: sets input file (default: stdin)
  {}[filename]: sets output file (default: stdout)
  {}[string]: sets ignored word to those contained in the string separated by spaces or comma
  \t(default: the builtin list that depends on the language)
  {}[string]: adds words contained in the string to the list of ignored words (default: none)
  {}[value]: sets max distance to be considered a repetition (in words) (default: 50)
  {}[value]: sets threshold value for underlining local repetitions (default: 1.9)
  {}[value|none]: activate global repetition detector and sets threshold value for underlining global repetitions
  \t(this threshold corresponds to the minimal ratio of words in the text, e.g. a threshold of 0.01 means
  \tthat a word must represent at least 1% of the total words in the text to be underlined) (default: not activated)
  {}[text|html]: sets input format (default: text, depends on input file extension)
  {}[terminal|html]|markdown]: sets output format (default: terminal, depends on output file extension)
  {}[true|false]: if true, try to detect proper nouns and don't count them (default: false)
  {}[value|none]: activate fuzzy string matching; value must be between 0.0 and 1.0 and corresponds to the maximal
  \t'difference' between two words until they are no more considered identical (e.g. 0.25 means that two words
  \t must have no more than 25% of difference) (default: not activated)",
             env!("CARGO_PKG_VERSION"),
             ARG_USAGE,
             ARG_VERSION,
             ARG_LIST_LANGUAGES,
             ARG_LANG,
             ARG_INPUT,
             ARG_OUTPUT,
             ARG_IGNORE,
             ARG_ADD_IGNORED,
             ARG_MAX_DISTANCE,
             ARG_THRESHOLD,
             ARG_GLOBAL_THRESHOLD,
             ARG_INPUT_FORMAT,
             ARG_OUTPUT_FORMAT,
             ARG_IGNORE_PROPER,
             ARG_FUZZY);
}

pub struct Config {
    pub lang: String,
    pub threshold: f32,
    pub global_threshold: Option<f32>,
    pub max_distance: u32,
    pub input_format: String,
    pub output_format: String,
    pub ignore_proper: bool,
    pub input: Box<Read>,
    pub input_filename: String,
    pub output: Box<Write>,
    pub output_filename: String,
    pub ignored: String,
    pub add_ignored: String,
    pub fuzzy: Option<f32>,
}

impl Config {
    /// New default config
    pub fn new() -> Config {
        Config {
            lang: "french".to_string(),
            threshold:1.9,
            global_threshold: None,
            max_distance:50,
            input_format: String::new(),
            output_format: String::new(),
            ignore_proper:false,
            input: Box::new(io::stdin()),
            input_filename: String::new(),
            output: Box::new(io::stdout()),
            output_filename: String::new(),
            ignored: String::new(),
            add_ignored: String::new(),
            fuzzy: None
        }
    }

    /// New config from program args
    pub fn new_from_args() -> Config {
        let mut config = Config::new();
        let mut iter = env::args();
        iter.next();
        // Sets fields from args
        for argument in iter {
            config.parse_arg(&argument);
        }
        // Sets fields to default values if they have not been set
        if config.input_format.is_empty() {
            if config.input_filename.ends_with(".html") {
                config.input_format = "html".to_string();
            } else {
                config.input_format = "text".to_string();
            }
        }
        if config.output_format.is_empty() {
            if config.output_filename.ends_with(".html") {
                config.output_format = "html".to_string();
            } else if config.output_filename.ends_with(".md") {
                config.output_format = "markdown".to_string();
            } else {
                config.output_format = "terminal".to_string();
            }
        }
        config
    }
    
    /// Parse a single argument
    pub fn parse_arg(&mut self, arg:&str) {
        if arg.starts_with(ARG_OUTPUT) {
            let option = &arg[ARG_OUTPUT.len()..];
            let result = File::create(option);
            match result {
                Ok(f) => {
                    self.output = Box::new(f);
                    self.output_filename = option.to_string();
                },
                Err(e) => {
                    println!("Error opening file {}: {}", option, e.description());
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_INPUT) {
            let option = &arg[ARG_INPUT.len()..];
            let result = File::open(option);
            match result {
                Ok(f) => {
                    self.input = Box::new(f);
                    self.input_filename = option.to_string();
                },
                Err(e) => {
                    println!("Error opening file {}: {}", option, e.description());
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_LANG) {
            let option = &arg[ARG_LANG.len()..];
            self.lang = option.to_string();
        } else if arg.starts_with(ARG_THRESHOLD) {
            let option = &arg[ARG_THRESHOLD.len()..];
            self.threshold = match option.parse() {
                Ok(x) => x,
                Err(_) => {
                    println!("Error passing argument to threshold: {}", option);
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_FUZZY) {
            let option = &arg[ARG_FUZZY.len()..];
            self.fuzzy = if option == "none" {
                None
            } else {
                match option.parse() {
                    Ok(x) => Some(x),
                    Err(_) => {
                        println!("Error passing argument to fuzzy: {}", option);
                        exit(0);
                    }
                }
            }
        } else if arg.starts_with(ARG_GLOBAL_THRESHOLD) {
            let option = &arg[ARG_GLOBAL_THRESHOLD.len()..];
            self.global_threshold = if option == "none" {
                None
            } else {
                match option.parse() {
                    Ok(x) => Some(x),
                    Err(_) => {
                        println!("Error passing argument to global threshold: {}", option);
                        exit(0);
                    }
                }
            }
        } else if arg.starts_with(ARG_MAX_DISTANCE) {
            let option = &arg[ARG_MAX_DISTANCE.len()..];
            self.max_distance = match option.parse() {
                Ok(x) => x,
                Err(_) => {
                    println!("Error passing argument to max_distance: {}", option);
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_INPUT_FORMAT) {
            let option = &arg[ARG_INPUT_FORMAT.len()..];
            match option {
                "html" => self.input_format = option.to_string(),
                "text" => self.input_format = option.to_string(),
                _ => {
                    println!("Wrong argument to {}: expected 'html' or 'text', received: {}", ARG_INPUT_FORMAT, option);
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_OUTPUT_FORMAT) {
            let option = &arg[ARG_OUTPUT_FORMAT.len()..];
            match option {
                "html" => self.input_format = option.to_string(),
                "terminal" => self.input_format = option.to_string(),
                "markdown" => self.input_format = option.to_string(),
                _ => {
                    println!("Wrong argument to {}: expected 'html', 'terminal' or 'markdown', received: {}", ARG_OUTPUT_FORMAT, option);
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_IGNORE_PROPER) {
            let option = &arg[ARG_IGNORE_PROPER.len()..];
            match option {
                "true" => self.ignore_proper = true,
                "false" => self.ignore_proper = false,
                _ => {
                    println!("Wrong argument to ignore_proper: expected 'true' or 'false', received: {}", option);
                    exit(0);
                }
            }
        } else if arg.starts_with(ARG_IGNORE) {
            let option = &arg[ARG_IGNORE.len()..];
            self.ignored = option.to_string();
        } else if arg.starts_with(ARG_ADD_IGNORED) {
            let option = &arg[ARG_ADD_IGNORED.len()..];
            self.add_ignored = option.to_string();
        } else if arg == ARG_USAGE {
            usage();
            exit(0);
        } else if arg == ARG_VERSION {
            version();
            exit(0);
        } else if arg == ARG_LIST_LANGUAGES {
            list_languages();
            exit(0);
        } else {
            println!("Unrecognized argument: {}. See {} for help", arg, ARG_USAGE);
            exit(0);
        }
    }
}
